src/database.cc

```diff
@@ -356,6 +361,131 @@ NAN_METHOD(Database::Configure) {
     NanReturnValue(args.This());
 }
 
+NAN_METHOD(Database::RegisterFunction) {
+    NanScope();
+    Database* db = ObjectWrap::Unwrap<Database>(args.This());
+
+    REQUIRE_ARGUMENTS(2);
+    REQUIRE_ARGUMENT_STRING(0, functionName);
+    REQUIRE_ARGUMENT_FUNCTION(1, callback);
+
+    FunctionBaton *baton = new FunctionBaton(db, *functionName, callback);
+    sqlite3_create_function(
+        db->_handle,
+        *functionName,
+        -1, // arbitrary number of args
+        SQLITE_UTF8 | SQLITE_DETERMINISTIC,
+        baton,
+        FunctionBeginExecute,
+        NULL,
+        NULL);
+
+    NanReturnValue(args.This());
+}
+
+void Database::FunctionBeginExecute(sqlite3_context *context, int argc, sqlite3_value **argv) {
+    FunctionBaton *fn = (FunctionBaton *)sqlite3_user_data(context);
+    ExecuteFunctionBaton *baton = new ExecuteFunctionBaton(fn, context, argc, argv);
+
+    // call to the main thread (uv_default_loop) for JS function to execute &
+    // wait for the result
+    uv_mutex_t mutex;
+    uv_mutex_init(&mutex);
+    uv_cond_init(&baton->condition);
+    uv_async_init(uv_default_loop(), &baton->async, (uv_async_cb)Database::AsyncFunctionExecute);
+    uv_async_send(&baton->async);
+    uv_mutex_lock(&mutex);
+    uv_cond_wait(&baton->condition, &mutex);
+    uv_mutex_unlock(&mutex);
+    uv_cond_destroy(&baton->condition);
+
+    delete baton;
+}
+
+
+void Database::AsyncFunctionExecute(uv_async_t *async) {
+    NanScope();
+
+    ExecuteFunctionBaton *baton = (ExecuteFunctionBaton *)async->data;
+    FunctionBaton *fn = baton->function;
+    Database *db = fn->db;
+    Local<Function> cb = NanNew(fn->callback);
+    sqlite3_context *context = baton->context;
+    sqlite3_value **values = baton->argv;
+    int argc = baton->argc;
+
+    if (!cb.IsEmpty() && cb->IsFunction()) {
+
+        // build the argument list for the function call
+        Local<Value> *argv = new Local<Value>[argc];
```

Comment:
Well, `vector` isn't really stack allocation; it's contents are on the heap as well. I'd like to use C++ features since they'll guarantee that objects will be cleaned up in case an exception is thrown. Likely not a big deal here since it won't be caught anyway, but just good coding practice in general.